### DISCUSSION TITLE
Should !! on record parameters add a null check in init setters?

### DISCUSSION THREAD

**Discussion by kimsey0** [dotnet/csharplang#5823] [Language Ideas]
The new `!!` parameter null-checking in C# 11 is useful for records, since it allows null-checking while using the concise parameter-based syntax:

```csharp
public record Person(string FirstName!!, string LastName!!);
```

This is roughly equivalent to:

```csharp
public record Person(string FirstName, string LastName) {
    public string FirstName { get; init; } = FirstName ?? throw new ArgumentNullException(nameof(FirstName));
    public string LastName { get; init; } = LastName ?? throw new ArgumentNullException(nameof(LastName));
}
```

However, as noted in #5652 and perhaps surprisingly for coming users of the `!!` syntax, that still allows assigning nulls to the properties via the `init` setters:

```csharp
var personWithoutLastName = new Person("FirstName", "LastName") with { LastName = null! };
```

Would it actually be less astonishing and more useful that `!!` generated null checks in the `init` setters as well? Something similar to:

```csharp
public record Person(string FirstName, string LastName)
{
    private readonly string _firstName = FirstName ?? throw new ArgumentNullException(nameof(FirstName));
    private readonly string _lastName = LastName ?? throw new ArgumentNullException(nameof(LastName));

    public string FirstName
    {
        get => _firstName;
        init => _firstName = value ?? throw new ArgumentNullException(nameof(value));
    }

    public string LastName
    {
        get => _lastName;
        init => _lastName = value ?? throw new ArgumentNullException(nameof(value));
    }
}
```

There might be concerns about the `value` name used in the exceptions thrown by the setter, as well as the fact that `!!` otherwise doesn't apply to setters, as described in:
https://github.com/dotnet/csharplang/blob/69033cfc115bff94b398bb6a6648f488ee823a27/proposals/param-nullchecking.md?plain=1#L282-L286

**Comment by jnm2** [comment]
I thought `value` was the only appropriate param name to use for argument exceptions in setters/initers. The BCL certainly has been this way.

**Reply by kimsey0** [reply]
I'm probably just il-informed. 🙂

**Reply by jnm2** [reply]
Sorry not you, I'm just commenting on [proposals/param-nullchecking.md](https://github.com/dotnet/csharplang/blob/69033cfc115bff94b398bb6a6648f488ee823a27/proposals/param-nullchecking.md?plain=1#L282-L286).

**Comment by HaloFour** [comment]
I also hope that the team reconsiders their decision here given that the parameter and the set/init accessor are both the same declaration.  Furthermore I think we should try to explore how to apply `!!` to property set/init accessors directly as well:

🍝 
```cs
public record C(string Foo!!) {
    public string Bar { get; set!!; }
}
```

**Reply by CyrusNajmabadi** [reply]
@HaloFour It definitely came up and we think it makes a lot of sense.  We left it out initially just to wait on feedback and to see how important it was comparatively.

**Reply by RikkiGibson** [reply]
I would also like to see it happen. It feels like the decision to "also null check in the synthesized setter/initter" should probably be made before shipping .NET 7.

**Reply by Eli-Black-Work** [reply]
@CyrusNajmabadi Any comment on making `!!` apply to `init`?

**Reply by CyrusNajmabadi** [reply]
My same comment above applies. 

> It definitely came up and we think it makes a lot of sense. We left it out initially just to wait on feedback and to see how important it was comparatively.

**Comment by AartBluestoke** [comment]
in the spirit of "we must guarantee this not be null" - it feels that a class should be able to enforce that other people can not set its members to null, otherwise, where do those guards/checks go: on each use of your own member without an unguarded setter? 

Either you need to guard the set , or you need to guard every use of the member.
If you guard the set without set!! you need to do an implementation that writes to a value field.
pit of success suggests that you don't force people to add fields to your class in order to guarantee that a class holds non-null values.

**Comment by sab39** [comment]
I agree that the current behavior is confusing, especially considering that `!!` in general refers to validation on parameters being passed, which would mean the setter/initter in this case. As well as being more intuitive the proposed alternative desugaring would seem to be more efficient, since it's entirely possible property getters will be called in hot paths, or at the very least be called an order of magnitude more times than the setter or initter. Performing the null check once on initialization seems to me like it would be noticeably more efficient than doing it every time the value is read.
